### ISSUE TITLE
Is there a need to provide negative samples for object detection

### ISSUE THREAD

**Issue by s0r2637** [ultralytics/yolov5#10784]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I wanted to know whether Yolov5 requires hard negative samples to be provided for training a object detection on a custom dataset.  Does Yolov5 automatically selects negative regions from the training images? Or do we need to provide hard negative samples?

**Comment by s0r2637** [comment]
I looked into the tutorials and I find that we can provide hard negative samples by including images with not objects and omitting the label for it in the label.txt file.  However, if I don't provide hard negative samples, does Yolov5 directly samples from the background of the positive samples to collect negative instances?

**Comment by JustasBart** [comment]
Hi, while I'm not 100% what yolov5 does with the samples exactly, I can definitely say with confidence that having negative samples is a really good idea as it it'll learn how to suppress making detection where there should be none. One example that I can give you is if you have a stationary camera and say you want it to learn how to detect some electrical components, if you're not going to perform this live then you can place your components and then run Inference, however if you were to just have a live stream of detections, what should happen when you expose your fingers to the model?

It's not like your fingers are a part of the background and or the thing that you want to detect it only becomes a partial part of the background if you stick your hand in there, but at that stage your model will likely to at least sometimes confuse your fingers with a class that it was trained on just because it simply hasn't seen a hand/fingers before...

However, if you give it either some negative data of your hands/fingers and/or at the very least give it some samples where your hand is visible it'll be a lot less likely to ever confuse that with anything else.

Therefore whatever your data is, simply try to optimise it where it'll never be surprised or "Caught off guard" if that means that you need to show it some negative data then by all means do...

Hope this helps... :rocket:

**Comment by s0r2637** [comment]
Thanks!  I will read it carefully and get back to you.  I appreciate your time.

**Comment by s0r2637** [comment]
How about we don't give any negative images?  Instead, the algorithm should take any single positive image and replace the pixels inside the ground truth bboxes with some random pixels and turn that positive image into a negative image.  Would that simple trick help?  I mean in Yolov5, there is a provision to provide images with no labels.  Let me know what you think.

**Comment by JustasBart** [comment]
@s0r2637 I'm mainly a C++ dev here so I haven't really had the chance to do any explorations through the guts of yolov5 or any other architecture per say, but my understanding of yolov5 is that it already treats the background as negative data such so that what you're describing is somewhat redundant although not entirely either(?).

My point was more so in the case of you don't always need negative data on it's own, but there are cases where it makes perfect sense to include some. Otherwise the model will learn to include the data from within the bounding boxes and the exclude any data outside of that.
But if you consider my example of a stationary camera looking at electronics components - where do human fingers play their role? Do they at all? Is it necessary?
Because those sort of questions you should be asking when it comes to negative data.

The other example that I could think of is say if you want to build a potato detection model but say every now and then it'll see an onion but being that it hasn't been exposed to any negative data of onions, it'll very likely detect that as a potato. And that's the other use case for negative data, to suppress similar looking things in order to distinguish them from one another. But of course that being said if both potatoes and onions are to be labelled as a part of the model anyway well then the negative data makes no sense.

In addition to this I've even heard of people in throwing in completely random negative data of just random crazy looking images from the Internet such as to make the Network work harder during draining in order to suppress any strange outliers that you wouldn't even really think of. Because remember that any AI ultimately doesn't really know/understand what it's looking at anyway, in fact it doesn't even know or think that it's 'looking' in the first place, it just happily crunches away the Maths in order to adjust the weights in order to magically make the model work, and so therefore it's very prune for taking shortcuts and the easiest way out, which in some cases leads to false negatives on similar looking yet not identical objects etc... To which again negative data can help it at least to a certain extend, but only where it's needed and where it makes sense.

So in conclusion you generally don't have to lose any sleep over negative data, but if you have a look at the cycle of any AI it's always **[Train -> Deploy -> Test]** and it always goes back to capture more data if needed and repeat the cycle again until you're happy - meaning that this is where you either capture more data in general or experiment with negative data or just keep on experimenting with multiple things throughout your developing stages until you get it right.

Anyway, good luck with your model! :rocket:

**Comment by hopeSerendipity** [comment]
> @s0r2637 I'm mainly a C++ dev here so I haven't really had the chance to do any explorations through the guts of yolov5 or any other architecture per say, but my understanding of yolov5 is that it already treats the background as negative data such so that what you're describing is somewhat redundant although not entirely either(?).
> 
> My point was more so in the case of you don't always need negative data on it's own, but there are cases where it makes perfect sense to include some. Otherwise the model will learn to include the data from within the bounding boxes and the exclude any data outside of that. But if you consider my example of a stationary camera looking at electronics components - where do human fingers play their role? Do they at all? Is it necessary? Because those sort of questions you should be asking when it comes to negative data.
> 
> The other example that I could think of is say if you want to build a potato detection model but say every now and then it'll see an onion but being that it hasn't been exposed to any negative data of onions, it'll very likely detect that as a potato. And that's the other use case for negative data, to suppress similar looking things in order to distinguish them from one another. But of course that being said if both potatoes and onions are to be labelled as a part of the model anyway well then the negative data makes no sense.
> 
> In addition to this I've even heard of people in throwing in completely random negative data of just random crazy looking images from the Internet such as to make the Network work harder during draining in order to suppress any strange outliers that you wouldn't even really think of. Because remember that any AI ultimately doesn't really know/understand what it's looking at anyway, in fact it doesn't even know or think that it's 'looking' in the first place, it just happily crunches away the Maths in order to adjust the weights in order to magically make the model work, and so therefore it's very prune for taking shortcuts and the easiest way out, which in some cases leads to false negatives on similar looking yet not identical objects etc... To which again negative data can help it at least to a certain extend, but only where it's needed and where it makes sense.
> 
> So in conclusion you generally don't have to lose any sleep over negative data, but if you have a look at the cycle of any AI it's always **[Train -> Deploy -> Test]** and it always goes back to capture more data if needed and repeat the cycle again until you're happy - meaning that this is where you either capture more data in general or experiment with negative data or just keep on experimenting with multiple things throughout your developing stages until you get it right.
> 
> Anyway, good luck with your model! 🚀

@JustasBart i want to know that how can i tell the model that image is negative sample.Should i just put the image in the dataset without label or their is other way  in order to mark something as negative sample.

**Comment by JustasBart** [comment]
@hopeSerendipity That's exactly it, anything that isn't a label comes negative data (Even entire unlabeled images).

Good luck! 🚀
